Feature/measured values

IvannikovG opened the pull request:
1). Openapi validation tests do not work yet, connection timeout
2). More tests can be written yet, at least for unit tests

MoritzStueber committed:
Make parameter definition reusable

MoritzStueber committed:
Add specification of resource representing measurement data at a station

MoritzStueber committed:
Also sketch CSV-representation of measurement values at a station

Commit:
Initialise handler

Commit:
Add more to new handler

Commit:
Correctly return voiConfigs and timeseriesData for multiple vois

Commit:
Test commit

Commit:
Implement JSON representation for measured-values

Commit:
Add rendering function for CSV representation

IvannikovG committed:
Fix bug making the page to crash in case of no quantities were passed

IvannikovG committed:
Fix bug returnunng not correct format for JSON representation if no query parameters were passed
	timeseries key was not present at all. Now will be [].

IvannikovG committed:
Move most of the representation and data-extracting function from handler into lib. Fix bug that caused crash if one of the VOI was 'unconfigured'

IvannikovG committed:
Add basic testing for measured-values endpoint hrlper functions

IvannikovG committed:
remove redundancy in returning empty timeseries instead of returnin none. Add test to verify compliance with openapi

IvannikovG committed:
Add more tests to measured-values-utils

IvannikovG committed:
Add one more test

MoritzStueber committed:
Run `standard --fix` to automatically fix code style issues

MoritzStueber committed:
Manually fix straightforward style issues the linter complained about

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -263,15 +264,22 @@ function getWeatherMosmix (WEATHER_DATA_BASE_PATH, voisConfigs) {
+    const voi = req.query.quantities
+    let vois = defaultParameter
+    if (voi) {
+      vois = voi.split(',')
```
Only works if there are more than one quantities, crashes for exactly one

IvannikovG replied:
Could not reproduce

IvannikovG replied:
Tested with 

http://localhost:5000/weather-stations/10505/measured-values?quantities=t_2m,aswdir_s,pmsl,t_2m&from=1597140000000&to=1597150000000

http://localhost:5000/weather-stations/10505/measured-values?quantities=t_2m&from=1597140000000&to=1597150000000

MoritzStueber replied:
Couldn't reproduce either :confused: ... Never mind

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -263,15 +264,22 @@ function getWeatherMosmix (WEATHER_DATA_BASE_PATH, voisConfigs) {
+    const sid = splitUrl[2]
 
     if (isNaN(startTimestamp)) {
       startTimestamp = moment.utc().subtract(25, 'hours').valueOf()
```
Should default to start/end of today in timezone `Europe/Berlin` instead, see OAS

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -281,72 +289,51 @@ function getWeatherReport (WEATHER_DATA_BASE_PATH, voisConfigs) {
+      res.status(500).send('received a request for REPORT for unconfigured VOI')
+      req.log.warn({ res: res }, 'received request for REPORT for unconfigured VOI')
+      return
+    }
```
* `500` is the wrong HTTP status code, this is a `400 Bad Request`
* should return response of type `application/problem+json` as indicated in the OAS (compare line 246 in `index.js`, possibly just call that)

IvannikovG replied:
Was not possible to call failValidation from index.js, produced errors.

Replaced with:
res.set('Content-Type', 'application/problem+json')
      res.status(400).json({
        title: 'Schema validation Failed',
        status: 400,
        detail: 'Received request for unconfigured voi',
      })

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -281,72 +289,51 @@ function getWeatherReport (WEATHER_DATA_BASE_PATH, voisConfigs) {
-        unit: voiConfig.target.unit,
-        data: timeseriesData
+      default: function () {
+        res.status(406).send('Not Acceptable')
```
Should return a problem detail object as suggested in RFC 7807, not plain text

* [ ] @MoritzStueber add to OAS
* [ ] @IvannikovG fix for all data handlers

IvannikovG replied:
Introduced a function in responce_utils: problemDetail that accepts res and config. Res is responce object and config is an object containing {title: ..., status: ...,  detail: ..} According to RFC 7807.

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -281,72 +289,51 @@ function getWeatherReport (WEATHER_DATA_BASE_PATH, voisConfigs) {
-      })
+    const timeseriesDataCollection = await csv.readTimeseriesDataReport(REPORT_DATA_BASE_PATH, startTimestamp, endTimestamp, sid)
+    // timeseriesDataArray is an array for timeseries for each voi: [[{}, {}, {}], [{}, {}, {}]]
+    // timeseriesDataArray = [[{str:int, str: null}, {...} <-tmstmp]<-voi,[{...}, {...}]<-voi]
```
Usually better to use comments to explain what is happening (possibly why) using natural language, not implementation details (also I don't get the notation in the second line). Example: `// Read all data for period, but only keep quantities requested by user` above the block

IvannikovG replied:
removed redundant comments

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -263,15 +264,22 @@ function getWeatherMosmix (WEATHER_DATA_BASE_PATH, voisConfigs) {
 
-// GET /weather/weather_reports/poi/:sid/:voi?startTimestamp=...&endTimestamp=...
-function getWeatherReport (WEATHER_DATA_BASE_PATH, voisConfigs) {
+// http://localhost:5000/weather-stations/10505/measured-values?quantities=t_2m&from=1597140000000&to=1597226400000
```
I guess you're copy-pasting this for testing interactively, but if we follow the pattern of the comments above the other handlers, this should be `GET /weather-stations/:stationId/measured-values?quantities=...&from=...&to=...`

IvannikovG replied:
True. Fixed

MoritzStueber commented on lib/measured_values_utils.js:
```diff
@@ -0,0 +1,141 @@
+const _ = require('lodash')
```
File header missing: license information, statement `use strict` (see other files)

IvannikovG replied:
Fixed

MoritzStueber commented on lib/measured_values_utils.js:
```diff
@@ -0,0 +1,141 @@
+
+// General functions
+
+function ensureVoiConfigsCorrectness (voiConfigs) {
```
Better name `checkValidityOfQuantityIds (voiConfigs) {}`? For me, 'ensure' implies that something broken is corrected and this is not what this function does. Also, `voiExists` instead of `trueVoisArray` might be more readable

IvannikovG replied:
Fixed

MoritzStueber commented on lib/measured_values_utils.js:
```diff
@@ -0,0 +1,141 @@
+  return trueVoisArray
+}
+
+function getTimeSeriesData (voiConfigs, timeseriesDataCollection) {
```
Better name `dropTimeseriesDataNotOfInterest (..) {}`? Didn't know what the function was about with the old name, too generic I think

IvannikovG replied:
Fixed

MoritzStueber commented on lib/measured_values_utils.js:
```diff
@@ -0,0 +1,141 @@
+  return timeSeriesData
+}
+
+function formatTimeseriesDataArray (timeseriesDataArray) {
```
* Better name `useSIunitsAndDropNaN () {}`?
* Please re-add the comments that existed, no reason to get rid of them!! See https://github.com/UdSAES/dwd_data_access/pull/4/files#diff-f6f98c8d039ad2cbafd38480ce2f132eL302-L323

IvannikovG replied:
Renamed and returned comments

MoritzStueber commented on lib/measured_values_utils.js:
```diff
@@ -0,0 +1,141 @@
+    description: `Quantities ${vois} measured at station ${sid}`,
+    data: []
+  }
+  for (let i = 0; i < voiConfigs.length; i += 1) {
```
Iteration based on numerical index assumes that `voiConfigs` and `timeseriesDataArray` have the same length and that their members are in the same order, can you guarantee that?

MoritzStueber commented on lib/measured_values_utils.js:
```diff
@@ -0,0 +1,141 @@
+}
+
+function getLastElementsFromTimeseries (timeseriesDataArray) {
+  // returns from each array in timeseriesDataArray first element.
```
Comment doesn't match code

IvannikovG replied:
Renamed this function to getHeadElementsFromTimeseries

MoritzStueber commented on lib/measured_values_utils.js:
```diff
@@ -0,0 +1,141 @@
+  return result
+}
+
+function getLastElementsFromTimeseries (timeseriesDataArray) {
```
Better name `getAllButFirstElementsFromTimeseries (..) {}`?

IvannikovG replied:
Renamed this function to getHeadElementsFromTimeSeries

MoritzStueber commented on lib/measured_values_utils.js:
```diff
@@ -0,0 +1,141 @@
+    return helper(getLastElementsFromTimeseries(timeseries), result)
+  }
+  return helper(timeseriesDataArray, [])
+}
```
Cool, recursion! :sunglasses: Hard to understand at first sight, though. Also, why was this necessary, what's the advantage over a simple `_.forEach()` or similar? Please add comments that explain&increase readability!

Also, if it's not really necessary, this adds complexity and thus should probably be implemented as simple as possible. Even if the resulting code is not that cool... Don't know though.

IvannikovG replied:
I would like to ask you to allow me to keep it. For each is not possible, because:
I do not exactly iterate through timeseries for each voi, but also kind of inside each voi too.
E.g. timeseries would be smth like:

[[voi1_timeseries_el1, voi1_timeseries_el2],
 [voi2_timeseries_el1, voi2_timeseries_el2],
 [voi3_timeseries_el1, voi3_timeseries_el2]]

And i iterate by columns kind of, getting [[voi1_timeseries_el1, voi2_timeseries_el1, ... voiN...]]

So in my opinion this is a good usecase for such recursion. We can refactor later on but for now i assume adding comments will be enough? Because it does what intended..

MoritzStueber replied:
All good, was just surprised and didn't really try to understand why it was necessary. Let's keep it :slightly_smiling_face:

MoritzStueber commented on lib/measured_values_utils.js:
```diff
@@ -0,0 +1,141 @@
+      return _.includes(timestampsToRemove, item.timestamp) && _.isNil(item.value)
+    })
+  })
+  return timeseriesDataArray
```
Unit conversion is missing! See https://github.com/UdSAES/dwd_data_access/pull/4/files#diff-f6f98c8d039ad2cbafd38480ce2f132eL326-L331

Here is probably a good place to implement this because it should happen before/outside the rendering functions.

IvannikovG replied:
Implemented in a separate function: convertUnits in utils

MoritzStueber commented on test/test_measured_values_utils.js:
```diff
@@ -0,0 +1,144 @@
+const _ = require('lodash')
+
+describe('Validate correctness of functions that manipulate VOIS', async function () {
+  const VOIS_DATA_ACCESS_CONFIGS_PATH = './config/vois_data_access.json'
```
Should probably go into a `before` or `beforeEach`-section to avoid code duplication and speed up execution

IvannikovG replied:
decided to keep same and rework in case of need when more tests are added to test/test_measured_values_utils.js.

MoritzStueber commented on test/test_measured_values_utils.js:
```diff
@@ -0,0 +1,144 @@
+    const vois = ['t_2m', 'pdml']
+    const voisDataAccessConfigs = getVoiConfigsAsArray(vois)
+    it('should return the expected output', async function () {
+      // Read expected result from .json-file
```
Comments no longer match code (throughout entire file) -- please update or remove; since the structure is always the same (expect, actual, assert) we can probably remove the comments

IvannikovG replied:
Fixed

MoritzStueber reviewed (changes requested):
Detailed feedback on the internals; didn't have a look at the API tests yet (probably tomorrow). Looks promising so far!

IvannikovG committed:
Make timestamps to default to the start and the end of the current date in Europe/Berlin timezone

IvannikovG committed:
Reword what is returned after trying to get uncofigured voi. 500 was replaced by 400 status

IvannikovG committed:
Rework the way 400 and 500 statuses are handeled.
	It is possible now to pass a config object that specifies problem ddetails according to RFC 7807 and
	Instead of returning plain text

IvannikovG committed:
Rename some functions, add Licencing to files

IvannikovG committed:
Rename function formatTimeseriesDataArray to useSIunitsAndDropNaN and add removed comments back there

IvannikovG committed:
Add comments to functions in lib

IvannikovG committed:
Add test to verify number of vois is the same as number of entries in timeseries data

IvannikovG committed:
Add test to validate there is as much timeseries as vois

IvannikovG committed:
Add missing unit conversion

MoritzStueber committed:
Install package `got`

MoritzStueber committed:
Replace implementation of API-tests for measured values

Hopefully sth. more robust that is somewhat maintainable. Replaced 
`axios` by `got` because axios doesn't support specifying a port for 
requests (WTF). Only works for 2xx 'cause `got.get()` throws otherwise. 
Check against OAS not yet implemented.

Commit:
Fix several linting issues

Commit:
Fix linting in data_access handler

MoritzStueber committed:
Merge branch 'feature/measured-values' of github.com:UdSAES/dwd_data_access into feature/measured-values

MoritzStueber committed:
Refactor function so that name matches purpose

MoritzStueber committed:
Clean up and comment source code

MoritzStueber committed:
Refactor test for improved conciseness/less duplicated code

MoritzStueber committed:
Get rid of non-helpful test

Test basically re-iterated an entire function definition and assertions 
were of the form "some value is the correct ballpark", to unspecific for 
my taste.

MoritzStueber committed:
Make test titles more specific/fix language issues

MoritzStueber committed:
Provide commands for running unit- and api-tests based on test name

API tests include `GET` -- not that clean, but it works for now.

MoritzStueber committed:
Remove trailing comma (=column) in CSV-representation

MoritzStueber committed:
Visually separate sections of source code by comments spanning line

MoritzStueber committed:
Make code for data selection more concise by increased usage of Lodash

MoritzStueber committed:
Replace non-robust implementation of unit conversion

Previously based on numerical indices, now relies on actual name-based 
lookups.

MoritzStueber committed:
Fix broken rendering function for JSON-representation

MoritzStueber committed:
Add log statements for debugging/tracing

MoritzStueber committed:
Fix broken unit test for rendering CSV-representation

MoritzStueber committed:
Make test heading more concise

MoritzStueber committed:
Fix code formatting/imprecise comments

MoritzStueber committed:
Add test harness for validating pre-processing of BEOB data (WIP)

Actual tests still missing, should be added eventually.

MoritzStueber committed:
Merge branch 'dev' into feature/measured-values

MoritzStueber committed:
Make test headings more descriptive
